File: test/problems/2-medium/238-Product-Of-Array-Except-Self/solution.cpp
#include <vector>

using namespace std;

vector<int> productExceptSelf(vector<int>& nums) {
    
    vector<int> prefix_product(nums.size());
    int current_product = 1;
    for (int i = 0; i < nums.size(); i++) {
        current_product *= nums[i];
        prefix_product[i] = current_product;
    }

    current_product = 1;
    vector<int> suffix_product(nums.size());
    for (int i = nums.size() - 1; i >= 0; i--) {
        current_product *= nums[i];
        suffix_product[i] = current_product;
    }

    vector<int> res(nums.size());
    for (int i = nums.size() - 1; i >= 0; i--) {
        if (i == 0) {
            res[i] = suffix_product[i+1];
        } else if (i == nums.size()-1) {
            res[i] = prefix_product[i-1];
        } else {
            res[i] = prefix_product[i-1]*suffix_product[i+1];
        }
    }

    return res;
}

File: test/problems/2-medium/238-Product-Of-Array-Except-Self/solution.java
public int[] productExceptSelf(int[] nums) {
    int n = nums.length;
    int[] prefixProduct = new int[n];
    int[] suffixProduct = new int[n];
    int[] res = new int[n];
    
    // Calculate prefix products
    int currentProduct = 1;
    for (int i = 0; i < n; i++) {
        currentProduct *= nums[i];
        prefixProduct[i] = currentProduct;
    }
    
    // Calculate suffix products
    currentProduct = 1;
    for (int i = n - 1; i >= 0; i--) {
        currentProduct *= nums[i];
        suffixProduct[i] = currentProduct;
    }
    
    // Calculate result
    for (int i = 0; i < n; i++) {
        if (i == 0) {
            res[i] = suffixProduct[i + 1];
        } else if (i == n - 1) {
            res[i] = prefixProduct[i - 1];
        } else {
            res[i] = prefixProduct[i - 1] * suffixProduct[i + 1];
        }
    }
    
    return res;
}